drain3/template_miner.py

```diff
@@ -170,24 +173,94 @@ def match(self, log_message: str, full_search_strategy="never") -> LogCluster:
+
+        :param log_template: log template corresponding to the log message
+        :param log_message: log message to extract parameters from
+        :param exact_matching: whether to apply the correct masking-patterns to match parameters, or try to approximate;
+            disabling exact_matching may be faster but may lead to situations in which parameters are wrongly identified
+        :return: An ordered list of parameter values present in the log message.
+        """
+        parameter_details_list = self.get_parameter_details_list(log_template, log_message, exact_matching=exact_matching)
+        return [parameter.value for parameter in parameter_details_list]
 
+    def get_parameter_details_list(self, log_template: str, log_message: str, exact_matching: bool = True) -> List[ParameterDetails]:
+        """
+        Extract parameters from a log message according to a provided template that was generated by Drain3 by calling `add_log_message()`.
+        This will work best if each `MaskingInstruction` has a `mask_with` unique across all instances
+        and none of them have a value of `*`.
+
+        :param log_template: log template corresponding to the log message
+        :param log_message: log message to extract parameters from
+        :param exact_matching: whether to apply the correct masking-patterns to match parameters, or try to approximate;
+            disabling exact_matching may be faster but may lead to situations in which parameters are wrongly identified
+        :return: A ordered list of ParameterDetails for the log message.
+        """
         for delimiter in self.config.drain_extra_delimiters:
-            content = re.sub(delimiter, ' ', content)
-        parameter_list = re.findall(template_regex, content)
-        parameter_list = parameter_list[0] if parameter_list else ()
-        parameter_list = list(parameter_list) if isinstance(parameter_list, tuple) else [parameter_list]
-
-        def is_mask(p: str):
-            return p.startswith(self.config.mask_prefix) and p.endswith(self.config.mask_suffix)
-
-        parameter_list = [p for p in list(parameter_list) if not is_mask(p)]
-        return parameter_list
+            log_message = re.sub(delimiter, " ", log_message)
+
+        template_regex, parameter_group_to_mask_name = self._template_to_regex(log_template, exact_matching)
+
+        # Parameters are represented by specific named groups inside template_regex.
+        parameter_match = re.match(template_regex, log_message)
+        parameter_details_list = []
+        if parameter_match:
+            for group_name, parameter in parameter_match.groupdict().items():
+                if group_name in parameter_group_to_mask_name:
+                    parameter_details_list.append(ParameterDetails(parameter, parameter_group_to_mask_name[group_name]))
+
+        return parameter_details_list
+
+    def _template_to_regex(self, log_template: str, exact_matching: bool):
```

Comment:
Lets call this `_get_template_parameter_extraction_regex()`

Reply:
I think caching here is a very good idea. 
We can use an  `LRUCache` instance defined in TemplateMiner instance.
A default size of 3000 should be reasonable.  
A config parameter in `masking` section can control it.